exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFunctionRegistry.java

```diff
@@ -64,62 +76,134 @@
         }
+
+        String[] names = holder.getRegisteredNames();
         for (String name : names) {
           String functionName = name.toLowerCase();
-          registeredFunctions.put(functionName, holder);
           String functionSignature = functionName + functionInput;
-          String existingImplementation;
-          if ((existingImplementation = functionSignatureMap.get(functionSignature)) != null) {
-            throw new AssertionError(
-                String.format(
-                    "Conflicting functions with similar signature found. Func Name: %s, Class name: %s " +
-                " Class name: %s", functionName, func.getClassName(), existingImplementation));
-          } else if (holder.isAggregating() && !holder.isDeterministic() ) {
-            logger.warn("Aggregate functions must be deterministic, did not register function {}", func.getClassName());
+
+          if (allFuncWithSignatures.get(functionName).contains(functionSignature)) {
+            throw new FunctionValidationException(
+                String.format("Conflicting function with similar signature found. " +
+                        "Function name: %s, class name: %s, input parameters : %s",
+                    functionName, func.getClassName(), functionInput));
+          } else if (holder.isAggregating() && !holder.isDeterministic()) {
+            throw new FunctionValidationException(
+                String.format("Aggregate functions must be deterministic: %s", func.getClassName()));
           } else {
-            functionSignatureMap.put(functionSignature, func.getClassName());
+            functions.add(Func.newBuilder().setName(functionName).addAllMajorType(types).build());
+            allFuncWithSignatures.put(functionName, functionSignature);
           }
         }
       } else {
         logger.warn("Unable to initialize function for class {}", func.getClassName());
       }
     }
-    if (logger.isTraceEnabled()) {
-      StringBuilder allFunctions = new StringBuilder();
-      for (DrillFuncHolder method: registeredFunctions.values()) {
-        allFunctions.append(method.toString()).append("\n");
+    return functions;
+  }
+
+  /**
+   * Registers all functions present in jar.
+   * If jar name is already registered, all jar related functions are overridden.
+   */
+  public void register(String jarName, ScanResult classpathScan, ClassLoader classloader) {
+    FunctionConverter converter = new FunctionConverter();
+    List<AnnotatedClassDescriptor> providerClasses = classpathScan.getAnnotatedClasses();
+    Map<String, Pair<String, DrillFuncHolder>> functions = Maps.newHashMap();
+    for (AnnotatedClassDescriptor func : providerClasses) {
+      DrillFuncHolder holder = converter.getHolder(func, classloader);
+      if (holder != null) {
+        String functionInput = "";
+        for (DrillFuncHolder.ValueReference ref : holder.parameters) {
+          functionInput += ref.getType().toString();
```

Comment:
String representation for function input is the following:
major_type {\n
minor_type: VARCHAR\n
mode: REQUIRED\n
}\n
major_type {\n
minor_type: INT\n
mode: REQUIRED\n
}\n
But I agree that since it's going to be part of public API, we definitely should think about better string representation.
functionInput contains only input types, functionSignature contains function name with functionInput.
So their string representation can be the following:
functionInput:
VARCHAR-REQUIRED, INT-REQUIRED
functionSignature:
MY_FUNC(VARCHAR-REQUIRED, INT-REQUIRED)